File: content/snippets/write-product-descriptions/n3.test.js
/**
 * Ces tests injectent un double local au lieu d'appeler un fournisseur.
 *
 * Ce qu'ils prouvent : le dossier arrive dans l'invite, la température demandée
 * est celle envoyée, un dossier trop grand est refusé avant toute dépense, une
 * panne est retentée, une réponse inutilisable lève, et une copie qui affirme
 * un attribut absent du dossier est refusée.
 *
 * Ce qu'ils ne prouvent pas : que le modèle écrit bien, ni qu'il écrit
 * autrement demain.
 */
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { FakeLLM } from '../_harness/fake-llm.mjs';
import { FakeSDK } from '../_harness/fake-sdk.mjs';
import essai from '../../tryouts/frozen/write-product-descriptions.js';
import {
  MAX_CHARACTERS,
  MIN_CHARACTERS,
  MODEL,
  DescriptionUnavailable,
  UngroundedDescription,
  describe,
  providerClient,
} from './n3.js';

const PRODUCT = {
  name: 'Aurore 500',
  category: 'sac à dos',
  material: 'toile recyclée',
  audience: 'les randonneurs',
  features: ['poche pour ordinateur', 'sangle ventrale'],
  colours: ['ardoise', 'sable'],
  warranty: 'deux ans',
};

const VOCABULARY = ['toile recyclée', 'cuir pleine fleur', 'étanche', 'poche pour ordinateur', 'garanti à vie'];

const COPY =
  'Aurore 500 tient la journée de marche sans se rappeler à vous. Sa toile ' +
  'recyclée encaisse les ronces, et sa poche pour ordinateur rentre au bureau le lundi.';

const FOR_LIFE =
  'Aurore 500 tient la journée de marche sans se rappeler à vous. Sa toile ' +
  'recyclée encaisse les ronces, et le sac est garanti à vie contre les défauts de couture.';

const answer = (description) => JSON.stringify({ description });

/** Les lignes d'attributs que l'extrait envoie pour une fiche de l'essai (valeurs sans virgule interne). */
const attributeText = (fiche) =>
  fiche
    .split('\n')
    .map((line) => {
      const [key, ...rest] = line.split(' : ');
      return `- ${key.trim()} : ${rest.join(' : ').trim()}`;
    })
    .join('\n');

const PROMPT = [
  "Tu rédiges la présentation d'un article pour une boutique en ligne.",
  "Écris deux phrases en français, sans superlatif, et n'affirme rien qui ne",
  'figure pas dans les caractéristiques ci-dessous.',
  'Réponds par un objet JSON et rien d\'autre : {"description": "…"}.',
  '',
  'Caractéristiques :',
].join('\n');

// ---------------------------------------------------------------------------
// Point de rupture
// ---------------------------------------------------------------------------

test('point de rupture : le modèle promet ce que la boutique ne vend pas', async () => {
  await assert.rejects(() => describe(PRODUCT, new FakeLLM({ response: answer(FOR_LIFE) }), { vocabulary: VOCABULARY }), (error) => {
    assert.ok(error instanceof UngroundedDescription);
    assert.equal(error.message, 'garanti à vie');
    return true;
  });
  const forLife = { ...PRODUCT, warranty: 'garanti à vie' };
  assert.equal(await describe(forLife, new FakeLLM({ response: answer(FOR_LIFE) }), { vocabulary: VOCABULARY }), FOR_LIFE);
});

test('point de rupture : et il ne vaut que ce que vaut votre liste', async () => {
  assert.equal(await describe(PRODUCT, new FakeLLM({ response: answer(FOR_LIFE) }), { vocabulary: ['étanche'] }), FOR_LIFE);
});

test('point de rupture : l’extrait demande une température non nulle', async () => {
  const client = new FakeLLM({ response: answer(COPY) });
  await describe(PRODUCT, client);
  assert.equal(client.lastRequest.temperature, 0.7);
  await describe(PRODUCT, client, { temperature: 0.4 });
  assert.equal(client.lastRequest.temperature, 0.4);
});

test('point de rupture : deux réponses différentes passent toutes deux le contrôle', async () => {
  // Ne démontre rien sur la variation du modèle : les deux copies sont écrites par le double.
  const first = await describe(PRODUCT, new FakeLLM({ response: answer(COPY) }), { vocabulary: VOCABULARY });
  const secondCopy =
    'Aurore 500 part en week-end sans y penser. Sa toile recyclée passe la pluie ' +
    'et les ronces, et son ardoise discrète se fait oublier en réunion.';
  const second = await describe(PRODUCT, new FakeLLM({ response: answer(secondCopy) }), { vocabulary: VOCABULARY });
  assert.notEqual(first, second);
});

test('point de rupture : l’essai montre la promesse dite autrement qui passe', async () => {
  const cas = essai.cases[5];
  assert.equal(cas.fails, true);
  const out = await essai.run(cas.input, 'fr', cas);
  assert.equal(out.verdict, undefined);
  assert.ok(out.output.includes('votre dos reste sec sous l’averse'));
  // Témoin : la promesse écrite avec le terme de la liste est refusée (cas 5).
  const refused = await essai.run(essai.cases[4].input, 'fr', essai.cases[4]);
  assert.equal(refused.verdict.label, 'Publication refusée : une promesse absente du dossier');
});

test('un dossier qui nie l’attribut l’ancre quand même', async () => {
  const record = { ...PRODUCT, features: [...PRODUCT.features, 'non étanche'] };
  const copy = 'Aurore 500 suit les randonneurs par tous les temps, sa toile recyclée est étanche.';
  await assert.rejects(() => describe(record, new FakeLLM({ response: answer(copy) }), { vocabulary: VOCABULARY }), UngroundedDescription);
});

test('un terme de la liste accordé passe le contrôle', async () => {
  const lamp = { ...PRODUCT, category: 'lampe de bureau', gender: 'f' };
  const copy = 'Aurore 500 est une lampe de bureau solide, garantie à vie contre les défauts.';
  await assert.rejects(() => describe(lamp, new FakeLLM({ response: answer(copy) }), { vocabulary: VOCABULARY }), UngroundedDescription);
});

// ---------------------------------------------------------------------------
// Docstring et commentaires
// ---------------------------------------------------------------------------

test('écrit la copie que le modèle a rendue', async () => {
  assert.equal(await describe(PRODUCT, new FakeLLM({ response: answer(COPY) }), { vocabulary: VOCABULARY }), COPY);
});

test('envoie des consignes en français et le dossier entier', async () => {
  const client = new FakeLLM({ response: answer(COPY) });
  await describe(PRODUCT, client);
  assert.equal(
    client.lastRequest.prompt,
    `${PROMPT}\n${[
      '- name : Aurore 500',
      '- category : sac à dos',
      '- material : toile recyclée',
      '- audience : les randonneurs',
      '- features : poche pour ordinateur, sangle ventrale',
      '- colours : ardoise, sable',
      '- warranty : deux ans',
    ].join('\n')}`,
  );
});

test('replie les retours à la ligne que le modèle laisse', async () => {
  assert.equal(await describe(PRODUCT, new FakeLLM({ response: answer(`  ${COPY}\n\n  `) })), COPY);
});

test('refuse un dossier trop grand avant de dépenser quoi que ce soit', async () => {
  const client = new FakeLLM({ response: answer(COPY) });
  await assert.rejects(() => describe({ name: 'Aurore 500', features: ['détail interminable '.repeat(40)] }, client), RangeError);
  assert.equal(client.callCount, 0);
  assert.equal(await describe({ n: 'x'.repeat(MAX_CHARACTERS - 6) }, client), COPY);
  await assert.rejects(() => describe({ n: 'x'.repeat(MAX_CHARACTERS - 5) }, client), RangeError);
  assert.equal(client.callCount, 1);
});

test('une panne est retentée le nombre de fois annoncé', async () => {
  let client = new FakeLLM({ response: answer(COPY), failTimes: 2 });
  assert.equal(await describe(PRODUCT, client, { attempts: 3 }), COPY);
  assert.equal(client.callCount, 3);
  client = new FakeLLM({ response: answer(COPY), failTimes: 5 });
  await assert.rejects(() => describe(PRODUCT, client), (error) => {
    assert.ok(error instanceof DescriptionUnavailable);
    assert.match(error.message, /simulated provider failure/);
    return true;
  });
  assert.equal(client.callCount, 3);
});

test('une réponse qui n’est pas du JSON lève plutôt que d’être publiée', async () => {
  const client = new FakeLLM({ response: 'Bien sûr ! Voici une proposition de description :' });
  await assert.rejects(() => describe(PRODUCT, client, { attempts: 2 }), DescriptionUnavailable);
  assert.equal(client.callCount, 2);
});

test('une réponse d’une autre forme lève', async () => {
  // JSON sans description, description vide, nombre, `null`, clôture non
  // refermée, prose autour d'une clôture.
  const responses = [
    JSON.stringify({ titre: 'Aurore 500' }),
    answer(''),
    answer(42),
    'null',
    `\`\`\`json\n${answer(COPY)}`,
    `Voici la description :\n\`\`\`json\n${answer(COPY)}\n\`\`\``,
  ];
  for (const response of responses) {
    await assert.rejects(() => describe(PRODUCT, new FakeLLM({ response }), { attempts: 1 }), DescriptionUnavailable);
  }
});

test('une description en liste est publiée, collée par une virgule', async () => {
  // Python lève DescriptionUnavailable sur la même réponse.
  const response = answer(['Aurore 500 tient la journée de marche.', 'Sa toile recyclée encaisse les ronces.']);
  await assert.rejects(() => describe(PRODUCT, new FakeLLM({ response }), { attempts: 1 }), DescriptionUnavailable);
});

test('un fragment est refusé', async () => {
  await assert.rejects(() => describe(PRODUCT, new FakeLLM({ response: answer('Un sac à dos.') })), DescriptionUnavailable);
  assert.equal(await describe(PRODUCT, new FakeLLM({ response: answer('A'.repeat(MIN_CHARACTERS)) })), 'A'.repeat(MIN_CHARACTERS));
  await assert.rejects(() => describe(PRODUCT, new FakeLLM({ response: answer('A'.repeat(MIN_CHARACTERS - 1)) })), DescriptionUnavailable);
});

test('la recherche ignore casse et accents', async () => {
  await assert.rejects(
    () => describe(PRODUCT, new FakeLLM({ response: answer(FOR_LIFE.replace('garanti à vie', 'GARANTI A VIE')) }), { vocabulary: VOCABULARY }),
    UngroundedDescription,
  );
  const record = { ...PRODUCT, warranty: 'Garanti À VIE'.normalize('NFD') };
  assert.equal(await describe(record, new FakeLLM({ response: answer(FOR_LIFE) }), { vocabulary: VOCABULARY }), FOR_LIFE);
});

test('le client est injecté pour tester sans réseau', async () => {
  await assert.rejects(() => describe(PRODUCT), (error) => {
    assert.equal(error.code, 'ERR_MODULE_NOT_FOUND');
    assert.match(error.message, /openai/);
    return true;
  });
});

test('production : l’adaptateur appelle la surface du vrai kit', async () => {
  // L'adaptateur sur le double du harnais, à la forme du kit `openai` publié,
  // sans méthode `complete` : `chat.completions.create({ model, messages,
  // temperature })`, réponse lue dans `choices[0].message.content`.
  const sdk = new FakeSDK({ content: answer(COPY) });
  assert.equal(sdk.complete, undefined);
  const client = await providerClient(sdk);
  assert.equal(await describe(PRODUCT, client, { temperature: 0.4 }), COPY);
  const { endpoint, model, messages, temperature } = sdk.lastRequest;
  assert.deepEqual([endpoint, model], ['chat.completions', MODEL]);
  assert.equal(messages[0].role, 'user');
  assert.ok(messages[0].content.includes(PRODUCT.name));
  // La température demandée est celle qui arrive au kit, pas une valeur par défaut.
  assert.equal(temperature, 0.4);
  assert.equal(sdk.requests.length, 1);
  const tiede = new FakeSDK({ content: answer(COPY) });
  await describe(PRODUCT, await providerClient(tiede));
  assert.equal(tiede.lastRequest.temperature, 0.7);
});

test('production : l’adaptateur, une réponse sans contenu lève après les essais', async () => {
  // Le kit type `content` comme facultatif : `null` n'est pas une copie.
  const sdk = new FakeSDK({ content: null });
  const client = await providerClient(sdk);
  await assert.rejects(() => describe(PRODUCT, client), DescriptionUnavailable);
  assert.equal(sdk.requests.length, 3);
});

test('production : l’adaptateur, une panne du kit est retentée', async () => {
  const sdk = new FakeSDK({ content: answer(COPY), failTimes: 2 });
  assert.equal(await describe(PRODUCT, await providerClient(sdk)), COPY);
  assert.equal(sdk.requests.length, 3);
  const mort = new FakeSDK({ content: answer(COPY), failTimes: 3 });
  const client = await providerClient(mort);
  await assert.rejects(() => describe(PRODUCT, client), DescriptionUnavailable);
  assert.equal(mort.requests.length, 3);
});

test('l’essai rend ses cinq premiers cas comme il les annonce', async () => {
  const run = (i, lang = 'fr') => essai.run(essai.cases[i].input, lang, essai.cases[i]);
  const first = await run(0);
  assert.equal(first.output, COPY);
  assert.equal(first.note, `1 appel facturé, ${PROMPT.length + 1 + attributeText(essai.cases[0].input).length} caractères partis chez le fournisseur, consignes et fiche comprises.`);
  assert.equal(first.note, '1 appel facturé, 490 caractères partis chez le fournisseur, consignes et fiche comprises.');
  const second = await run(1);
  assert.equal(second.verdict.label, 'Refusé avant le premier appel');
  assert.equal(second.note, 'Rien n’est parti chez le fournisseur, rien n’est dû.');
  assert.equal((await run(2)).note, '3 appels facturés : les deux premiers ont échoué, le troisième a répondu.');
  const prose = await run(3);
  assert.equal(prose.verdict.label, 'Réponse rejetée');
  assert.equal(prose.note, '3 appels envoyés et facturés, aucune réponse publiable.');
  assert.equal((await run(4)).verdict.detail, 'Le contrôle terme à terme a trouvé « garanti à vie » dans le texte et nulle part dans le dossier.');
});

test('l’essai compte bien 702 caractères d’attributs dans la fiche encombrée', async () => {
  // `shown` : « 702 caractères d’attributs, 102 de trop ».
  assert.equal(attributeText(essai.cases[1].input).length, 702);
  assert.equal(702 - MAX_CHARACTERS, 102);
  assert.ok(essai.cases[1].shown.fr.includes('702 caractères d’attributs, 102 de trop'));
});

// ---------------------------------------------------------------------------
// Cas de production
// ---------------------------------------------------------------------------

test('production : un dossier vide ne part pas chez le fournisseur', async () => {
  // Un dossier sans un attribut ne peut rien donner : il est refusé avant de
  // coûter un appel.
  const client = new FakeLLM({ response: answer(COPY) });
  await assert.rejects(() => describe({}, client), { name: 'RangeError', message: /empty product record/ });
  assert.equal(client.callCount, 0);
});

test('production : zéro essai lève sans appel', async () => {
  const client = new FakeLLM({ response: answer(COPY) });
  await assert.rejects(() => describe(PRODUCT, client, { attempts: 0 }), DescriptionUnavailable);
  assert.equal(client.callCount, 0);
});

test('le plafond compte des unités UTF-16', async () => {
  // 296 emojis : 305 caractères d'attributs en Python, acceptés ; 601 unités UTF-16 ici.
  let out;
  try {
    out = await describe({ name: '🙂'.repeat(296) }, new FakeLLM({ response: answer(COPY) }));
  } catch (error) {
    assert.fail(`${error.name}: ${error.message}`);
  }
  assert.equal(out, COPY);
});

File: content/snippets/write-product-descriptions/n3.test.py
"""
Ces tests injectent un double local au lieu d'appeler un fournisseur.

Ce qu'ils prouvent : le dossier arrive dans l'invite, la température demandée
est celle envoyée, un dossier trop grand est refusé avant toute dépense, une
panne est retentée, une réponse inutilisable lève, et une copie qui affirme un
attribut absent du dossier est refusée.

Ce qu'ils ne prouvent pas : que le modèle écrit bien, ni qu'il écrit autrement
demain.
"""

import json
import unicodedata

import pytest

from _harness.fake_llm import FakeLLM
from _harness.fake_sdk import FakeSDK
from n3 import (
    MAX_CHARACTERS,
    MIN_CHARACTERS,
    MODEL,
    PROMPT,
    DescriptionUnavailable,
    ProviderClient,
    UngroundedDescription,
    describe,
)

PRODUCT = {
    "name": "Aurore 500",
    "category": "sac à dos",
    "material": "toile recyclée",
    "audience": "les randonneurs",
    "features": ["poche pour ordinateur", "sangle ventrale"],
    "colours": ["ardoise", "sable"],
    "warranty": "deux ans",
}

VOCABULARY = ("toile recyclée", "cuir pleine fleur", "étanche", "poche pour ordinateur", "garanti à vie")

COPY = (
    "Aurore 500 tient la journée de marche sans se rappeler à vous. Sa toile "
    "recyclée encaisse les ronces, et sa poche pour ordinateur rentre au bureau le lundi."
)

FOR_LIFE = (
    "Aurore 500 tient la journée de marche sans se rappeler à vous. Sa toile "
    "recyclée encaisse les ronces, et le sac est garanti à vie contre les défauts de couture."
)


def answer(description) -> str:
    return json.dumps({"description": description})


# ---------------------------------------------------------------------------
# Point de rupture
# ---------------------------------------------------------------------------


def test_point_de_rupture_le_modele_promet_ce_que_la_boutique_ne_vend_pas():
    """
    « Le test fait écrire au double local que le sac est « garanti à vie »
    quand le dossier dit deux ans […] Ce qui l'attrape est le contrôle du
    niveau du dessous, terme à terme contre le dossier » (existant).
    """
    with pytest.raises(UngroundedDescription) as refused:
        describe(PRODUCT, FakeLLM(response=answer(FOR_LIFE)), vocabulary=VOCABULARY)
    assert str(refused.value) == "garanti à vie"
    # Témoin : pour un dossier dont la garantie est réellement à vie, la phrase passe.
    for_life = {**PRODUCT, "warranty": "garanti à vie"}
    assert describe(for_life, FakeLLM(response=answer(FOR_LIFE)), vocabulary=VOCABULARY) == FOR_LIFE


def test_point_de_rupture_et_il_ne_vaut_que_ce_que_vaut_votre_liste():
    """« et il ne vaut que ce que vaut votre liste » : sans « garanti à vie » dans la liste, la promesse passe."""
    assert describe(PRODUCT, FakeLLM(response=answer(FOR_LIFE)), vocabulary=("étanche",)) == FOR_LIFE


def test_point_de_rupture_l_extrait_demande_une_temperature_non_nulle():
    """« l'extrait demande une température non nulle » : 0,7 par défaut, et celle de l'appelant sinon (existant, complété)."""
    client = FakeLLM(response=answer(COPY))
    describe(PRODUCT, client)
    assert client.last_request["temperature"] == 0.7
    describe(PRODUCT, client, temperature=0.4)
    assert client.last_request["temperature"] == 0.4


def test_point_de_rupture_deux_reponses_differentes_passent_toutes_deux_le_controle():
    """
    « deux appels sur le même produit ne rendent pas le même texte » : non
    testable sur le modèle. Ce test (existant) montre seulement que deux copies
    différentes, écrites par le double, passent toutes deux : il ne démontre
    rien sur la variation du modèle. Voir le relevé.
    """
    first = describe(PRODUCT, FakeLLM(response=answer(COPY)), vocabulary=VOCABULARY)
    second_copy = (
        "Aurore 500 part en week-end sans y penser. Sa toile recyclée passe la pluie "
        "et les ronces, et son ardoise discrète se fait oublier en réunion."
    )
    second = describe(PRODUCT, FakeLLM(response=answer(second_copy)), vocabulary=VOCABULARY)
    assert first != second


def test_defaut_un_dossier_qui_nie_l_attribut_ne_l_ancre_pas():
    record = {**PRODUCT, "features": [*PRODUCT["features"], "non étanche"]}
    copy = "Aurore 500 suit les randonneurs par tous les temps, sa toile recyclée est étanche."
    with pytest.raises(UngroundedDescription):
        describe(record, FakeLLM(response=answer(copy)), vocabulary=VOCABULARY)


def test_defaut_un_terme_de_la_liste_accorde_est_refuse_aussi():
    lamp = {**PRODUCT, "category": "lampe de bureau", "gender": "f"}
    copy = "Aurore 500 est une lampe de bureau solide, garantie à vie contre les défauts."
    with pytest.raises(UngroundedDescription):
        describe(lamp, FakeLLM(response=answer(copy)), vocabulary=VOCABULARY)


# ---------------------------------------------------------------------------
# Docstring et commentaires
# ---------------------------------------------------------------------------


def test_ecrit_la_copie_que_le_modele_a_rendue():
    """Cas nominal (existant)."""
    assert describe(PRODUCT, FakeLLM(response=answer(COPY)), vocabulary=VOCABULARY) == COPY


def test_envoie_des_consignes_en_francais_et_le_dossier_entier():
    """
    « the request » ; commentaire « The instructions are written in the language
    of the shop » ; risks `data_egress: third-party` (existant, resserré : invite exacte).
    """
    client = FakeLLM(response=answer(COPY))
    describe(PRODUCT, client)
    assert client.last_request["prompt"] == PROMPT + "\n" + "\n".join(
        [
            "- name : Aurore 500",
            "- category : sac à dos",
            "- material : toile recyclée",
            "- audience : les randonneurs",
            "- features : poche pour ordinateur, sangle ventrale",
            "- colours : ardoise, sable",
            "- warranty : deux ans",
        ]
    )
    assert PROMPT.startswith("Tu rédiges la présentation d'un article")


def test_replie_les_retours_a_la_ligne_que_le_modele_laisse():
    """Existant."""
    assert describe(PRODUCT, FakeLLM(response=answer(f"  {COPY}\n\n  "))) == COPY


def test_refuse_un_dossier_trop_grand_avant_de_depenser_quoi_que_ce_soit():
    """Commentaire : « Refusing an oversized record is not an optimisation, it is a cost control » (existant, complété : limites)."""
    client = FakeLLM(response=answer(COPY))
    with pytest.raises(ValueError):
        describe({"name": "Aurore 500", "features": ["détail interminable " * 40]}, client)
    assert client.call_count == 0
    assert describe({"n": "x" * (MAX_CHARACTERS - 6)}, client) == COPY
    with pytest.raises(ValueError):
        describe({"n": "x" * (MAX_CHARACTERS - 5)}, client)
    assert client.call_count == 1


def test_une_panne_est_retentee_le_nombre_de_fois_annonce():
    """« the retry » (existant, complété)."""
    client = FakeLLM(response=answer(COPY), fail_times=2)
    assert describe(PRODUCT, client, attempts=3) == COPY
    assert client.call_count == 3
    client = FakeLLM(response=answer(COPY), fail_times=5)
    with pytest.raises(DescriptionUnavailable, match="simulated provider failure"):
        describe(PRODUCT, client)
    assert client.call_count == 3


def test_une_reponse_qui_n_est_pas_du_json_leve_plutot_que_d_etre_publiee():
    """« an answer that is only probably JSON » (existant) : chaque essai est un appel facturé."""
    client = FakeLLM(response="Bien sûr ! Voici une proposition de description :")
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, client, attempts=2)
    assert client.call_count == 2


def test_une_reponse_d_une_autre_forme_leve():
    """JSON sans description, description vide, nombre, liste, `null`, clôture non refermée, prose autour d'une clôture."""
    for response in (
        json.dumps({"titre": "Aurore 500"}),
        answer(""),
        answer(42),
        answer(["Aurore 500 tient la journée de marche.", "Sa toile recyclée encaisse les ronces."]),
        "null",
        "```json\n" + answer(COPY),
        "Voici la description :\n```json\n" + answer(COPY) + "\n```",
    ):
        with pytest.raises(DescriptionUnavailable):
            describe(PRODUCT, FakeLLM(response=response), attempts=1)


def test_un_fragment_est_refuse():
    """`MIN_CHARACTERS` (existant, complété : limites)."""
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, FakeLLM(response=answer("Un sac à dos.")))
    assert describe(PRODUCT, FakeLLM(response=answer("A" * MIN_CHARACTERS))) == "A" * MIN_CHARACTERS
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, FakeLLM(response=answer("A" * (MIN_CHARACTERS - 1))))


def test_la_recherche_ignore_casse_et_accents():
    """`_fold` : « so « À vie » meets « a vie » »."""
    with pytest.raises(UngroundedDescription):
        describe(PRODUCT, FakeLLM(response=answer(FOR_LIFE.replace("garanti à vie", "GARANTI A VIE"))), vocabulary=VOCABULARY)
    record = {**PRODUCT, "warranty": unicodedata.normalize("NFD", "Garanti À VIE")}
    assert describe(record, FakeLLM(response=answer(FOR_LIFE)), vocabulary=VOCABULARY) == FOR_LIFE


def test_le_client_est_injecte_pour_tester_sans_reseau():
    """« `client` is injected so this can be tested without a network call »."""
    with pytest.raises(ModuleNotFoundError, match="openai"):
        describe(PRODUCT)


def test_production_l_adaptateur_appelle_la_surface_du_vrai_kit():
    """
    docstring de `ProviderClient` : « The one call this snippet makes, on top of
    the provider's SDK ». Sur le double du harnais, à la forme du kit `openai`
    publié, sans méthode `complete` : `chat.completions.create(model=...,
    messages=[...], temperature=...)`, réponse lue dans
    `choices[0].message.content`.
    """
    sdk = FakeSDK(content=answer(COPY))
    assert not hasattr(sdk, "complete")
    assert describe(PRODUCT, ProviderClient(sdk=sdk), temperature=0.4) == COPY
    request = sdk.last_request
    assert request["endpoint"] == "chat.completions"
    assert request["model"] == MODEL == "gpt-4.1-mini"
    assert request["messages"][0]["role"] == "user"
    assert PRODUCT["name"] in request["messages"][0]["content"]
    # La température demandée est celle qui arrive au kit, pas une valeur par défaut.
    assert request["temperature"] == 0.4
    assert len(sdk.requests) == 1
    tiede = FakeSDK(content=answer(COPY))
    describe(PRODUCT, ProviderClient(sdk=tiede))
    assert tiede.last_request["temperature"] == 0.7


def test_production_l_adaptateur_une_reponse_sans_contenu_leve_apres_les_essais():
    """Le kit type `content` comme facultatif : `None` n'est pas une copie, et il est redemandé."""
    sdk = FakeSDK(content=None)
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, ProviderClient(sdk=sdk))
    assert len(sdk.requests) == 3


def test_production_l_adaptateur_une_panne_du_kit_est_retentee():
    sdk = FakeSDK(content=answer(COPY), fail_times=2)
    assert describe(PRODUCT, ProviderClient(sdk=sdk)) == COPY
    assert len(sdk.requests) == 3
    mort = FakeSDK(content=answer(COPY), fail_times=3)
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, ProviderClient(sdk=mort))
    assert len(mort.requests) == 3


# ---------------------------------------------------------------------------
# Cas de production
# ---------------------------------------------------------------------------


def test_production_un_dossier_vide_ne_part_pas_chez_le_fournisseur():
    """Un dossier sans un attribut ne peut rien donner : il est refusé avant de coûter un appel."""
    client = FakeLLM(response=answer(COPY))
    with pytest.raises(ValueError, match="empty product record"):
        describe({}, client)
    assert client.call_count == 0


def test_production_zero_essai_leve_sans_appel():
    client = FakeLLM(response=answer(COPY))
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, client, attempts=0)
    assert client.call_count == 0


def test_production_le_plafond_compte_des_caracteres():
    """Python : 296 emojis font 305 caractères d'attributs, acceptés (JavaScript les refuse, voir n3.test.js)."""
    assert describe({"name": "🙂" * 296}, FakeLLM(response=answer(COPY))) == COPY
